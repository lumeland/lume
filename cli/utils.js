import { existsSync } from "../deps/fs.ts";
import lume, { overrides } from "../mod.js";
import { join, resolve, toFileUrl } from "../deps/path.ts";
import { bold, brightGreen, dim, red } from "../deps/colors.ts";

/**
 * Create and configure a site instance
 *
 * @param {Object} options user options to override
 * @return {Promise<*>} a lume instance - ready to build, run, etc.
 */
export async function buildSite(options) {
  options.root = resolve(Deno.cwd(), options.root);
  const configFile = join(options.root, options.config);

  overrides.cwd = options.root;

  if (options.dev) {
    overrides.dev = options.dev;
  }

  if (options.location) {
    overrides.location = new URL(options.location);
  }

  if (options.src) {
    overrides.src = options.src;
  }

  if (options.dest) {
    overrides.dest = options.dest;
  }

  if (options.port) {
    (overrides.server ||= {}).port = parseInt(options.port);
  }

  if (options.open) {
    (overrides.server ||= {}).open = options.open;
  }

  if (options["--"]) {
    overrides.flags = options["--"];
  }

  if (existsSync(configFile)) {
    const mod = await import(toFileUrl(configFile));
    return mod.default;
  }

  return lume();
}

/**
 * Validate the number of arguments passed to a command
 *
 * @param {string} command name of the command you're validating
 * @param {[]} args array of cli arguments (i.e. options._)
 * @param {number} max the max number of args
 * @param {number} min the min number of args
 */
export function validateArgsCount(command, args, max, min = 0) {
  if (args.length > max) {
    throw new Error(`Unexpected arguments: ${args.slice(max).join(", ")}

    Run ${brightGreen(`lume ${command} --help`)} for usage information
    `);
  }

  if (args.length < min) {
    throw new Error(`Missing arguments

    Run ${brightGreen(`lume ${command} --help`)} for usage information
    `);
  }
}

export function getCurrentVersion() {
  const url = new URL("../", import.meta.url).pathname;
  return url.match(/@([^/]+)/)?.[1] ?? `local (${url})`;
}

export function printError(exception, debug = false) {
  console.log();
  console.error(bold(red(`Error:`)), exception.message);

  printDataError(exception);

  if (debug) {
    console.error(exception);
  }

  if (exception.error) {
    let { error } = exception;
    let indent = "  ";
    while (error) {
      console.log();
      console.log(red(`${indent}${error.message}`));
      printDataError(error, indent);
      indent += "  ";

      if (debug) {
        console.error(error);
      }

      error = error.error;
    }
  }

  console.log();
}

function printDataError(error, indent = "") {
  if (!error.data) {
    return;
  }

  for (let [key, value] of Object.entries(error.data)) {
    if (key === "page") {
      value = value.src.path + value.src.ext;
    }

    console.log(dim(`${indent}${key}:`), value);
  }
}
