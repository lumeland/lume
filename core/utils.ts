import { DOMParser, HTMLDocument } from "../deps/dom.ts";
import { SEP } from "../deps/path.ts";

/** Run a callback concurrently with all the elements of an Iterable */
export async function concurrent<Type>(
  iterable: AsyncIterable<Type> | Iterable<Type>,
  iteratorFn: (arg: Type) => Promise<unknown>,
  limit = 200,
) {
  const executing: Promise<unknown>[] = [];

  for await (const item of iterable) {
    const p: Promise<unknown> = iteratorFn(item).then(() =>
      executing.splice(executing.indexOf(p), 1)
    );

    executing.push(p);

    if (executing.length >= limit) {
      await Promise.race(executing);
    }
  }

  await Promise.all(executing);
}

/**
 * The list of supported MIME types.
 * It's used by the server and some plugins.
 */
export const mimes: Map<string, string> = new Map([
  [".apng", "image/apng"],
  [".atom", "application/atom+xml; charset=utf-8"],
  [".avif", "image/avif"],
  [".css", "text/css; charset=utf-8"],
  [".gif", "image/gif"],
  [".html", "text/html; charset=utf-8"],
  [".ico", "image/x-icon"],
  [".jpeg", "image/jpg"],
  [".jpg", "image/jpg"],
  [".js", "text/javascript; charset=utf-8"],
  [".json", "application/json"],
  [".map", "application/json"],
  [".mp3", "audio/mpeg"],
  [".mp4", "video/mp4"],
  [".ogg", "audio/ogg"],
  [".otf", "font/otf"],
  [".pdf", "application/pdf"],
  [".png", "image/png"],
  [".rss", "application/rss+xml; charset=utf-8"],
  [".svg", "image/svg+xml; charset=utf-8"],
  [".ttf", "font/ttf"],
  [".txt", "text/plain; charset=utf-8"],
  [".wasm", "application/wasm"],
  [".webm", "video/webm"],
  [".webmanifest", "application/manifest+json"],
  [".webp", "image/webp"],
  [".woff", "font/woff"],
  [".woff2", "font/woff2"],
  [".xml", "text/xml"],
  [".zip", "application/zip"],
]);

/**
 * Merge two objects recursively.
 * It's used to merge user options with default options.
 */
export function merge<Type>(
  defaults: Type,
  user?: Partial<Type>,
) {
  const merged = { ...defaults };

  if (!user) {
    return merged;
  }

  for (const [key, value] of Object.entries(user)) {
    // @ts-ignore: No index signature with a parameter of type 'string' was found on type 'unknown'
    if (isPlainObject(merged[key]) && isPlainObject(value)) {
      // @ts-ignore: Type 'string' cannot be used to index type 'Type'
      merged[key] = merge(merged[key], value);
      continue;
    }

    // @ts-ignore: Type 'string' cannot be used to index type 'Type'
    merged[key] = value;
  }

  return merged;
}

/** Check if the argument passed is a plain object */
function isPlainObject(obj: unknown) {
  return typeof obj === "object" && obj !== null &&
    obj.toString() === "[object Object]";
}

/**
 * Convert the Windows paths (that use the separator "\")
 * to Posix paths (with the separator "/").
 */
export function normalizePath(path: string) {
  return SEP === "/" ? path : path.replaceAll(SEP, "/");
}

/**
 * Search an extension in a map.
 * It's useful for cases in which the extension is multiple.
 * Example: page.tmpl.js
 */
export function searchByExtension<Type>(
  path: string,
  extensions: Map<string, Type>,
): [string, Type] | undefined {
  for (const [key, value] of extensions) {
    if (path.endsWith(key)) {
      return [key, value];
    }
  }
}

/** Convert an HTMLDocument instance to a string */
export function documentToString(document: HTMLDocument) {
  const { doctype, documentElement } = document;

  if (!doctype) {
    return documentElement?.outerHTML || "";
  }

  return `<!DOCTYPE ${doctype.name}` +
    (doctype.publicId ? ` PUBLIC "${doctype.publicId}"` : "") +
    (!doctype.publicId && doctype.systemId ? " SYSTEM" : "") +
    (doctype.systemId ? ` "${doctype.systemId}"` : "") +
    `>\n${documentElement?.outerHTML}`;
}

const parser = new DOMParser();

/** Parse a string with HTML code and return an HTMLDocument */
export function stringToDocument(string: string): HTMLDocument {
  const document = parser.parseFromString(string, "text/html");

  if (!document) {
    throw new Error("Unable to parse the HTML code");
  }

  return document;
}

/**
 * Generic Exception to throw errors.
 * It allows to include extra data and the previous exception.
 */
export class Exception extends Error {
  data?: Record<string, unknown>;
  error?: Error;

  constructor(message: string, data?: Record<string, unknown>, error?: Error) {
    super(message);
    this.data = data;
    this.error = error;
  }
}
