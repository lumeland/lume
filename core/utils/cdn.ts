import {
  canParse,
  format,
  maxSatisfying,
  parse,
  parseRange,
} from "../../deps/semver.ts";
import { fromFileUrl, globToRegExp, toFileUrl } from "../../deps/path.ts";
import { walk, type WalkOptions } from "../../deps/fs.ts";

type PackageType = "npm" | "gh";

export function getFiles(
  specifier: string,
  patterns: string[] = ["/**"],
): Promise<Map<string, string>> {
  if (specifier.startsWith("file:")) {
    return getFsFiles(specifier, patterns);
  }
  return getCDNFiles(specifier, patterns);
}

async function getCDNFiles(
  specifier: string,
  patterns: string[],
): Promise<Map<string, string>> {
  const result = parseNpm(specifier) || parseGh(specifier) ||
    parseJsDelivr(specifier);
  if (!result) {
    throw new Error(`Invalid specifier: ${specifier}`);
  }
  const [type, name, range, path] = result;
  const version = await getVersion(type, name, range);
  if (!version) {
    throw new Error(`No matching version found for ${name}@${range}`);
  }
  const url =
    `https://data.jsdelivr.com/v1/package/${type}/${name}@${version}?structure=flat`;
  const res = await fetch(url);
  if (!res.ok) {
    throw new Error(`Failed to fetch package data for ${name}@${version}`);
  }
  const data = await res.json();
  const regexps = patterns.map((pattern) =>
    globToRegExp(path + pattern, { globstar: true, extended: true })
  );
  const fileMap = new Map<string, string>();

  for (const file of data.files) {
    const filename: string = file.name;

    if (!regexps.some((regexp) => regexp.test(filename))) {
      continue;
    }

    const key = path ? filename.slice(path.length) : filename;
    fileMap.set(
      key,
      `https://cdn.jsdelivr.net/${type}/${name}@${version}${filename}`,
    );
  }

  return fileMap;
}

async function getFsFiles(
  specifier: string,
  patterns: string[],
): Promise<Map<string, string>> {
  const basePath = fromFileUrl(specifier);
  const fileMap = new Map<string, string>();
  const walkOptions: WalkOptions = {
    includeDirs: false,
    followSymlinks: false,
    skip: [
      /(^|\/)\.[^\/\.]/, // hidden files
    ],
  };

  const regexps = patterns.map((pattern) =>
    globToRegExp(pattern, { globstar: true, extended: true })
  );

  for await (const { path } of walk(basePath, walkOptions)) {
    const filename = path.slice(basePath.length);
    if (!regexps.some((regexp) => regexp.test(filename))) {
      continue;
    }
    fileMap.set(filename, toFileUrl(path).href);
  }

  return fileMap;
}

export function isFromCdn(specifier: string): boolean {
  return specifier.startsWith("npm:") ||
    specifier.startsWith("gh:") ||
    specifier.startsWith("https://cdn.jsdelivr.net/npm/") ||
    specifier.startsWith("https://cdn.jsdelivr.net/gh/");
}

export function getFile(specifier: string): string {
  if (specifier.startsWith("npm:")) {
    return specifier.replace("npm:", "https://cdn.jsdelivr.net/npm/");
  } else if (specifier.startsWith("gh:")) {
    return specifier.replace("gh:", "https://cdn.jsdelivr.net/gh/");
  } else if (
    specifier.startsWith("https://cdn.jsdelivr.net/npm/") ||
    specifier.startsWith("https://cdn.jsdelivr.net/gh/")
  ) {
    return specifier;
  } else {
    throw new Error(`Invalid specifier: ${specifier}`);
  }
}

export async function getVersion(
  type: PackageType,
  name: string,
  version: string,
): Promise<string | undefined> {
  const url = `https://data.jsdelivr.com/v1/package/${type}/${name}`;
  const res = await fetch(url);
  if (!res.ok) {
    throw new Error(`Failed to fetch latest version for ${name}`);
  }
  const data = await res.json();

  const versions = data.versions.filter(canParse).map(parse);
  const range = parseRange(version);
  const found = maxSatisfying(versions, range);
  return found ? format(found) : undefined;
}

const NPM = /^npm:(@[^/]+\/)?([^@/]+)(@([^/]+))?(\/.+)?$/;
function parseNpm(
  specifier: string,
): [PackageType, string, string, string] | undefined {
  const match = specifier.match(NPM);

  if (!match) {
    return;
  }

  const [, scope, name, , version, path] = match;
  return [
    "npm",
    (scope || "") + name,
    version || "*",
    path || "",
  ];
}

const GITHUB = /^gh:([^/]+\/[^/@]+)(@([^/]+))?(\/.+)?$/;
function parseGh(
  specifier: string,
): [PackageType, string, string, string] | undefined {
  const match = specifier.match(GITHUB);

  if (!match) {
    return;
  }

  const [, name, , version, path] = match;
  return [
    "gh",
    name,
    version || "*",
    path || "",
  ];
}

function parseJsDelivr(specifier: string): [
  PackageType,
  string,
  string,
  string,
] | undefined {
  if (specifier.startsWith("https://cdn.jsdelivr.net/npm/")) {
    return parseNpm(
      specifier.replace("https://cdn.jsdelivr.net/npm/", "npm:"),
    );
  }
  if (specifier.startsWith("https://cdn.jsdelivr.net/gh/")) {
    return parseGh(
      specifier.replace("https://cdn.jsdelivr.net/gh/", "gh:"),
    );
  }
}
