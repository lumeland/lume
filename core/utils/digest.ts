import { crypto } from "../../deps/crypto.ts";
import { encodeHex } from "../../deps/hex.ts";

const decoder = new TextDecoder();
const encoder = new TextEncoder();

/** Digest a message using SHA-1 algorithm */
export async function sha1(
  message: string | Uint8Array<ArrayBuffer>,
): Promise<string> {
  if (typeof message === "string") {
    message = encoder.encode(message);
  }

  const hash = await crypto.subtle.digest("SHA-1", message);
  return decoder.decode(hash);
}

/** Digest a message using MD5 algorithm */
export async function md5(
  message: string | Uint8Array<ArrayBuffer>,
): Promise<string> {
  if (typeof message === "string") {
    message = encoder.encode(message);
  }

  const hash = await crypto.subtle.digest("MD5", message);
  return encodeHex(hash);
}
