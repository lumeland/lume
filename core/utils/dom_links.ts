export interface Link {
  element: Element;
  attribute: string;
  value: string;
}

const selectors = {
  href: "[href]",
  src: "[src]",
  poster: "video[poster]",
  srcset: "[srcset]",
  imagesrcset: "[imagesrcset]",
  action: "form[action]",
};

export function* searchLinks(document: Document): Iterable<Link> {
  for (const [attribute, selector] of Object.entries(selectors)) {
    for (const element of document.querySelectorAll(selector)) {
      yield {
        element,
        attribute,
        value: element.getAttribute(attribute)?.trim() ?? "",
      };
    }
  }
}

export function parseSrcset(srcset: string): [string, string][] {
  return srcset.trim().split(",").map((src) => {
    const [, url, rest] = src.trim().match(/^(\S+)(.*)/)!;
    return [url, rest];
  });
}
