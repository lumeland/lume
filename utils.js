import { DOMParser } from "./deps/dom.ts";
import { SEP } from "./deps/path.ts";

export async function concurrent(iterable, iteratorFn, limit = 200) {
  const executing = [];

  for await (const item of iterable) {
    const p = iteratorFn(item).then(() =>
      executing.splice(executing.indexOf(p), 1)
    );

    executing.push(p);

    if (executing.length >= limit) {
      await Promise.race(executing);
    }
  }

  await Promise.all(executing);
}

export const mimes = new Map([
  [".apng", "image/apng"],
  [".atom", "application/atom+xml; charset=utf-8"],
  [".avif", "image/avif"],
  [".css", "text/css; charset=utf-8"],
  [".gif", "image/gif"],
  [".html", "text/html; charset=utf-8"],
  [".ico", "image/x-icon"],
  [".jpeg", "image/jpg"],
  [".jpg", "image/jpg"],
  [".js", "text/javascript; charset=utf-8"],
  [".json", "application/json"],
  [".map", "application/json"],
  [".mp3", "audio/mpeg"],
  [".mp4", "video/mp4"],
  [".ogg", "audio/ogg"],
  [".otf", "font/otf"],
  [".pdf", "application/pdf"],
  [".png", "image/png"],
  [".rss", "application/rss+xml; charset=utf-8"],
  [".svg", "image/svg+xml; charset=utf-8"],
  [".ttf", "font/ttf"],
  [".txt", "text/plain; charset=utf-8"],
  [".wasm", "application/wasm"],
  [".webm", "video/webm"],
  [".webmanifest", "application/manifest+json"],
  [".webp", "image/webp"],
  [".woff", "font/woff"],
  [".woff2", "font/woff2"],
  [".xml", "text/xml"],
  [".zip", "application/zip"],
]);

export function merge(defaults, user) {
  const merged = { ...defaults };

  if (!user) {
    return merged;
  }

  for (const [key, value] of Object.entries(user)) {
    if (isPlainObject(merged[key]) && isPlainObject(value)) {
      merged[key] = merge(merged[key], value);
      continue;
    }

    merged[key] = value;
  }

  return merged;
}

function isPlainObject(obj) {
  return typeof obj === "object" && obj.toString() === "[object Object]";
}

export function normalizePath(path) {
  return SEP === "/" ? path : path.replaceAll(SEP, "/");
}

export function searchByExtension(path, extensions) {
  for (const [key, value] of extensions) {
    if (path.endsWith(key)) {
      return [key, value];
    }
  }
}

export function documentToString(document) {
  const { doctype } = document;

  return `<!DOCTYPE ${doctype.name}` +
    (doctype.publicId ? ` PUBLIC "${doctype.publicId}"` : "") +
    (!doctype.publicId && doctype.systemId ? " SYSTEM" : "") +
    (doctype.systemId ? ` "${doctype.systemId}"` : "") +
    `>\n${document.documentElement.outerHTML}`;
}

const parser = new DOMParser();

export function stringToDocument(string) {
  return parser.parseFromString(string, "text/html");
}

export class Exception extends Error {
  constructor(message, data, error) {
    super(message);
    this.data = data;
    this.error = error;
  }
}
