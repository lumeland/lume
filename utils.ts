import { DOMParser, HTMLDocument } from "./deps/dom.ts";
import { SEP } from "./deps/path.ts";

/**
 * Helper to run concurrently a function for all elements of a Iterable.
 */
export async function concurrent<Type>(
  iterable: AsyncIterable<Type> | Iterable<Type>,
  iteratorFn: (a: Type) => Promise<void>,
  limit = 200,
): Promise<void> {
  const executing: unknown[] = [];

  for await (const item of iterable) {
    const p: unknown = iteratorFn(item).then(() =>
      executing.splice(executing.indexOf(p), 1)
    );

    executing.push(p);

    if (executing.length >= limit) {
      await Promise.race(executing);
    }
  }

  await Promise.all(executing);
}

/**
 * The list of supported mime types.
 * It's used by the server and some plugins.
 */
export const mimes = new Map([
  [".apng", "image/apng"],
  [".atom", "application/atom+xml; charset=utf-8"],
  [".avif", "image/avif"],
  [".css", "text/css; charset=utf-8"],
  [".gif", "image/gif"],
  [".html", "text/html; charset=utf-8"],
  [".ico", "image/x-icon"],
  [".jpeg", "image/jpg"],
  [".jpg", "image/jpg"],
  [".js", "text/javascript; charset=utf-8"],
  [".json", "application/json"],
  [".map", "application/json"],
  [".mp3", "audio/mpeg"],
  [".mp4", "video/mp4"],
  [".ogg", "audio/ogg"],
  [".otf", "font/otf"],
  [".pdf", "application/pdf"],
  [".png", "image/png"],
  [".rss", "application/rss+xml; charset=utf-8"],
  [".svg", "image/svg+xml; charset=utf-8"],
  [".ttf", "font/ttf"],
  [".txt", "text/plain; charset=utf-8"],
  [".wasm", "application/wasm"],
  [".webm", "video/webm"],
  [".webmanifest", "application/manifest+json"],
  [".webp", "image/webp"],
  [".woff", "font/woff"],
  [".woff2", "font/woff2"],
  [".xml", "text/xml"],
  [".zip", "application/zip"],
]);

/**
 * Merge two objects recursively.
 * It's used to merge user options with default options
 */
export function merge<Type>(
  defaults: Type,
  user: Record<string, unknown>,
): Type {
  const merged = { ...defaults };

  if (!user) {
    return merged;
  }

  for (const [key, value] of Object.entries(user)) {
    // @ts-ignore: I don't know how to type this
    if (isPlainObject(merged[key]) && isPlainObject(value)) {
      // @ts-ignore: I don't know how to type this
      merged[key] = merge(merged[key], value);
      continue;
    }

    // @ts-ignore: I don't know how to type this
    merged[key] = value;
  }

  return merged;
}

/**
 * Returns true if the argument passed is a plain object.
 */
function isPlainObject(obj: unknown): boolean {
  return typeof obj === "object" && obj !== null &&
    obj.toString() === "[object Object]";
}

/**
 * Convert the Windows paths (that use the separator "\")
 * to Posix paths (with the separator "/").
 */
export function normalizePath(path: string): string {
  return SEP === "/" ? path : path.replaceAll(SEP, "/");
}

/**
 * Search an extension in a Map.
 * It's useful for cases in which the extension is multiple.
 * Example: page.tmpl.js
 */
export function searchByExtension<Type>(
  path: string,
  extensions: Map<string, Type>,
): undefined | [string, Type] {
  for (const [key, value] of extensions) {
    if (path.endsWith(key)) {
      return [key, value];
    }
  }
}

/**
 * Converts a HTMLDocument instance to string code.
 */
export function documentToString(document: HTMLDocument): string {
  if (!document.documentElement) {
    return "";
  }

  const { doctype } = document;

  if (!doctype) {
    return document.documentElement.outerHTML;
  }

  return `<!DOCTYPE ${doctype.name}` +
    (doctype.publicId ? ` PUBLIC "${doctype.publicId}"` : "") +
    (!doctype.publicId && doctype.systemId ? " SYSTEM" : "") +
    (doctype.systemId ? ` "${doctype.systemId}"` : "") +
    `>\n${document.documentElement.outerHTML}`;
}

const parser = new DOMParser();

/**
 * Parse a string with HTML code and return a HTMLDocument
 */
export function stringToDocument(string: string): HTMLDocument | null {
  return parser.parseFromString(string, "text/html");
}

/**
 * Generic Exception to throw errors.
 * It allows to include extra data and the previous exception.
 */
export class Exception extends Error {
  data: unknown;
  error?: Error;

  constructor(message: string, data: unknown, error?: Error) {
    super(message);
    this.data = data;
    this.error = error;
  }
}
