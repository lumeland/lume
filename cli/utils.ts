import { Page, Site } from "../core.ts";
import lume from "../mod.ts";
import { exists } from "../deps/fs.ts";
import { join, resolve, toFileUrl } from "../deps/path.ts";
import { bold, dim, red } from "../deps/colors.ts";
import { Exception } from "../core/utils.ts";

/** Return the current installed version */
export function getCurrentVersion(): string {
  const url = new URL("../", import.meta.url).pathname;
  return url.match(/@([^/]+)/)?.[1] ?? `local (${url})`;
}

/** Return the latest stable version from the deno.land/x repository */
export async function getLastVersion(): Promise<string> {
  const response = await fetch("https://cdn.deno.land/lume/meta/versions.json");
  const versions = await response.json();
  return versions.latest;
}

/** Return the hash of the latest commit from the GitHub repository */
export async function getLastDevelopmentVersion(): Promise<string> {
  const response = await fetch(
    "https://api.github.com/repos/lumeland/lume/commits?per_page=1",
  );
  const commits = await response.json();
  return commits[0].sha;
}

/** Create a site instance */
export async function createSite(root: string, config?: string): Promise<Site> {
  root = resolve(Deno.cwd(), root);

  if (config) {
    const path = join(root, config);

    if (await exists(path)) {
      console.log(`Loading config file ${dim(config)}`);
      console.log();
      const mod = await import(toFileUrl(path).href);
      return mod.default;
    }

    throw new Exception("Config file not found", { path });
  }

  const files = ["_config.js", "_config.ts"];

  for (const file of files) {
    const path = join(root, file);

    if (await exists(path)) {
      console.log(`Loading config file ${dim(file)}`);
      console.log();
      const mod = await import(toFileUrl(path).href);
      return mod.default;
    }
  }

  return lume();
}

/** Pretty-print an Error or Exception instance */
export function printError(
  error: Error | Exception,
  indent = 0,
  stackLines = 1,
) {
  console.log();
  const tab = "  ".repeat(indent);

  console.error(`${tab}${bold(red(`Error:`))}`, error.message);

  if (error instanceof Exception) {
    for (let [key, value] of Object.entries(error.data ?? {})) {
      if (key === "page") {
        value = (value as Page).src.path + (value as Page).src.ext;
      }
      console.log(dim(`${tab}${key}:`), value);
    }
  }

  if (error.stack) {
    const stack = error.stack.split("\n");

    // Skip all the stack lines that have been already presented
    stack.slice(1, stack.length - stackLines).forEach((line) => {
      console.log(`${tab}${line.trim()}`);
      stackLines++;
    });
  }

  if (error instanceof Exception && error.error) {
    printError(error.error, indent + 1, stackLines);
  }

  if (indent == 0) {
    console.log();
  }
}

/** A list of the available optional plugins */
export const pluginNames = [
  "attributes",
  "base_path",
  "bundler",
  "code_highlight",
  "date",
  "eta",
  "inline",
  "jsx",
  "postcss",
  "pug",
  "relative_urls",
  "slugify_urls",
  "svg",
  "terser",
];
