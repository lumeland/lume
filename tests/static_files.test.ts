import { assertSiteSnapshot, build, getSite } from "./utils.ts";

Deno.test("Copy static files", async (t) => {
  const site = getSite({
    src: "static_files",
  });

  site.copy("static", ".");
  site.copy("other");
  site.copy(
    "posts/2022-01-01_first-post/assets",
    (file) => file.replace(".scss", ".css"),
  );
  site.copy("posts/2022-01-01_first-post/individual-file");
  site.copy([".css", ".js"]);
  site.copy([".copy"]);
  site.copy("_headers");
  site.copy("static/_redirects", "_redirects");
  site.copy(
    [".copy2"],
    (file) => "/subdir" + file.replace(/\.copy2/, ".copy3"),
  );

  // a single file can be copied multiple times
  site.copy("one.yes");
  site.copy("one.yes", "one-again.yes");

  // copied with the trailing slash
  site.copy("other2/");

  // not copied because of the trailing slash
  site.copy("three.no/");

  await build(site);
  await assertSiteSnapshot(t, site);
});
