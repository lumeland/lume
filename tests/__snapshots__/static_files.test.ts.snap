export const snapshot = {};

snapshot[`Copy static files 1`] = `
{
  formats: [
    {
      engines: 1,
      ext: ".tmpl.ts",
      pageLoader: [AsyncFunction: module],
    },
    {
      engines: 1,
      ext: ".tmpl.js",
      pageLoader: [AsyncFunction: module],
    },
    {
      engines: undefined,
      ext: ".tmpl.json",
      pageLoader: [AsyncFunction: json],
    },
    {
      dataLoader: [AsyncFunction: json],
      engines: undefined,
      ext: ".json",
    },
    {
      engines: 1,
      ext: ".md",
      pageLoader: [AsyncFunction: text],
    },
    {
      engines: 1,
      ext: ".markdown",
      pageLoader: [AsyncFunction: text],
    },
    {
      componentLoader: [AsyncFunction: module],
      copy: true,
      dataLoader: [AsyncFunction: module],
      engines: 1,
      ext: ".js",
    },
    {
      componentLoader: [AsyncFunction: module],
      dataLoader: [AsyncFunction: module],
      engines: 1,
      ext: ".ts",
    },
    {
      componentLoader: [AsyncFunction: text],
      engines: 1,
      ext: ".njk",
      includesPath: "_includes",
      pageLoader: [AsyncFunction: text],
    },
    {
      dataLoader: [AsyncFunction: yaml],
      engines: undefined,
      ext: ".yaml",
      pageLoader: [AsyncFunction: yaml],
    },
    {
      dataLoader: [AsyncFunction: yaml],
      engines: undefined,
      ext: ".yml",
      pageLoader: [AsyncFunction: yaml],
    },
    {
      copy: true,
      engines: undefined,
      ext: ".css",
    },
    {
      copy: true,
      engines: undefined,
      ext: ".copy",
    },
    {
      copy: [Function (anonymous)],
      engines: undefined,
      ext: ".copy2",
    },
  ],
}
`;

snapshot[`Copy static files 2`] = `
[
  {
    entry: "/_headers",
    flags: [],
    outputPath: "/_headers",
  },
  {
    entry: "/one.yes",
    flags: [],
    outputPath: "/one.yes",
  },
  {
    entry: "/one.yes",
    flags: [],
    outputPath: "/one-again.yes",
  },
  {
    entry: "/other/one",
    flags: [],
    outputPath: "/other/one",
  },
  {
    entry: "/other/two",
    flags: [],
    outputPath: "/other/two",
  },
  {
    entry: "/other2/one",
    flags: [],
    outputPath: "/other2/one",
  },
  {
    entry: "/other2/two",
    flags: [],
    outputPath: "/other2/two",
  },
  {
    entry: "/posts/2022-01-01_first-post/assets/inner/inner2/styles.scss",
    flags: [],
    outputPath: "/posts/first-post/assets/inner/inner2/styles.css",
  },
  {
    entry: "/posts/2022-01-01_first-post/assets/inner/styles1.scss",
    flags: [],
    outputPath: "/posts/first-post/assets/inner/styles1.css",
  },
  {
    entry: "/posts/2022-01-01_first-post/assets/styles.scss",
    flags: [],
    outputPath: "/posts/first-post/assets/styles.css",
  },
  {
    entry: "/posts/2022-01-01_first-post/callback.copy2",
    flags: [],
    outputPath: "/subdir/posts/first-post/callback.copy3",
  },
  {
    entry: "/posts/2022-01-01_first-post/individual-file",
    flags: [],
    outputPath: "/posts/first-post/individual-file",
  },
  {
    entry: "/posts/2022-01-01_first-post/inner/callback1.copy2",
    flags: [],
    outputPath: "/subdir/posts/first-post/inner/callback1.copy3",
  },
  {
    entry: "/posts/2022-01-01_first-post/inner/inner2/callback2.copy2",
    flags: [],
    outputPath: "/subdir/posts/first-post/inner/inner2/callback2.copy3",
  },
  {
    entry: "/posts/2022-01-01_first-post/inner/inner2/to-copy2.copy",
    flags: [],
    outputPath: "/posts/first-post/inner/inner2/to-copy2.copy",
  },
  {
    entry: "/posts/2022-01-01_first-post/inner/to-copy1.copy",
    flags: [],
    outputPath: "/posts/first-post/inner/to-copy1.copy",
  },
  {
    entry: "/posts/2022-01-01_first-post/to-copy.copy",
    flags: [],
    outputPath: "/posts/first-post/to-copy.copy",
  },
  {
    entry: "/script/2022-01-01_app/main.js",
    flags: [],
    outputPath: "/script/app/main.js",
  },
  {
    entry: "/static/_redirects",
    flags: [],
    outputPath: "/_redirects",
  },
  {
    entry: "/static/one.yes",
    flags: [],
    outputPath: "/one.yes",
  },
  {
    entry: "/static/two.yes",
    flags: [],
    outputPath: "/two.yes",
  },
  {
    entry: "/styles.css",
    flags: [],
    outputPath: "/styles.css",
  },
  {
    entry: "/styles/main.css",
    flags: [],
    outputPath: "/styles/main.css",
  },
]
`;

snapshot[`Copy static files 3`] = `[]`;
